### PR TITLE
Event stats

### PR DESCRIPTION
**Pull request by isa-leroux448**
### Changes
- added bar charts, pie charts and stats tables to event stats page according to [Figma](https://www.figma.com/design/CNLGXpCQswX2i5LgKZp4eV/Web-App-UI-Design---BizTech-2024?node-id=314-4759&t=TDGulqKcRTFNgfrI-1)
- created a percentage bar chart
- updated mock DB data type to match data currently stored in our registrations database
- edited stats table to work with same input data type as bar chart and pie chart 

### Out Of Scope
- backend connection

### Notes
- I had some issues with the react google charts responsiveness where they wouldn't scale when I resized the screen. I added a function that reloads the charts on screen resize for both the pie chart and bar chart
- I don't think we store school value in our registration data so I omitted that chart
- Also had some styling issues related to conflicts between tailwind and inline styling using dynamic widths for the chart boxes so I ended up using a CSS module for simplicity

### Image
Desktop1
Desktop2

https://github.com/user-attachments/assets/398dcb4c-f33f-4a7c-9f2d-210870750b86

### CODE REVIEW

**Comment by isa-leroux448** [comment]
I ran into a slight type issue. I created a new type called "AttendeeDetailedInformation" based on the registration data in the DB. I edited the return value of the fetchRegistrationData function to use this type. However, I realize that we are using quite a few different types (e.g. type Attendee) to represent attendee data and I'm not sure if they should all use the same type? It's causing a type error in the check but I'm hesitant to change any types and mess up another part of the code. 
Any recommendations? Should we consolidate the different types into one?

File: src/components/Stats/PercentageBars.tsx [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+const calculateBarPercentages = (
+  data: StatsChartData[],
+  total: number
+) => {
+  const percentages: { [key: string]: number } = {};
+
+  data.forEach((item) => {
+    const { label, value } = item;
+    percentages[label] = (value / total) * 100;
```
Comment (AllanT102) [inline src/components/Stats/PercentageBars.tsx:18]: is total always gonna be non zero?

File: src/types/types.ts [hunk 2]
```diff
@@ -70,4 +70,21 @@ export type AttendeeBasicInformation = BasicInformation & {
     diet: string;
     heardFrom: string;
     faculty: string;
-  };
\ No newline at end of file
+  };
+
+export type AttendeeDetailedInformation = {
```
Comment (AllanT102) [inline src/types/types.ts:75]: Types should be consistent with db. So i think this here should be "Registration"

**Comment by AllanT102** [comment]
> I ran into a slight type issue. I created a new type called "AttendeeDetailedInformation" based on the registration data in the DB. I edited the return value of the fetchRegistrationData function to use this type. However, I realize that we are using quite a few different types (e.g. type Attendee) to represent attendee data and I'm not sure if they should all use the same type? It's causing a type error in the check but I'm hesitant to change any types and mess up another part of the code. Any recommendations? Should we consolidate the different types into one?

It's best to be consistent yes. I did a quick look and it should be fine using AttendeeDetailedInformation if that represents the registration information. currently fetchRegistrationData is used for the registration table on admin side and the stats page, so it won't be a large change.

Please change this field to be "Registration" so that it matches the db table name. 

Also, the fetchBackend function is on dev now and it works. Please implement the call to the backend and fetch real data.

**Comment by voctory** [comment]
> > I ran into a slight type issue. I created a new type called "AttendeeDetailedInformation" based on the registration data in the DB. I edited the return value of the fetchRegistrationData function to use this type. However, I realize that we are using quite a few different types (e.g. type Attendee) to represent attendee data and I'm not sure if they should all use the same type? It's causing a type error in the check but I'm hesitant to change any types and mess up another part of the code. Any recommendations? Should we consolidate the different types into one?
> 
> It's best to be consistent yes. I did a quick look and it should be fine using AttendeeDetailedInformation if that represents the registration information. currently fetchRegistrationData is used for the registration table on admin side and the stats page, so it won't be a large change.
> 
> Please change this field to be "Registration" so that it matches the db table name.
> 
> Also, the fetchBackend function is on dev now and it works. Please implement the call to the backend and fetch real data.
> 
> @isa-leroux448

Hey @AllanT102, do we need to get the backend functionality working in this PR? It might be better to merge our efforts to fetch event and registrations data on both the Stats page and the Admin Events Registrations page, otherwise we're duplicating the amount of work to get both pages working.

**Comment by AllanT102** [comment]
> > > I ran into a slight type issue. I created a new type called "AttendeeDetailedInformation" based on the registration data in the DB. I edited the return value of the fetchRegistrationData function to use this type. However, I realize that we are using quite a few different types (e.g. type Attendee) to represent attendee data and I'm not sure if they should all use the same type? It's causing a type error in the check but I'm hesitant to change any types and mess up another part of the code. Any recommendations? Should we consolidate the different types into one?
> > 
> > 
> > It's best to be consistent yes. I did a quick look and it should be fine using AttendeeDetailedInformation if that represents the registration information. currently fetchRegistrationData is used for the registration table on admin side and the stats page, so it won't be a large change.
> > Please change this field to be "Registration" so that it matches the db table name.
> > Also, the fetchBackend function is on dev now and it works. Please implement the call to the backend and fetch real data.
> > @isa-leroux448
> 
> Hey @AllanT102, do we need to get the backend functionality working in this PR? It might be better to merge our efforts to fetch event and registrations data on both the Stats page and the Admin Events Registrations page, otherwise we're duplicating the amount of work to get both pages working.

True... ok let's do that @isa-leroux448 can you handle this then? Once you merge this in, I've assigned you the ticket to connect backend for registration table and stats

**Comment by isa-leroux448** [comment]
I fixed the userInfo component used in the attendee information popup to use the new Registration type and be able to handle the basicInformation object. Right now I'm running into issues when building with next but I don't think they're related to my branch. I think some broken stuff got merged into dev because the package.lock syntax is off and I get a whole bunch of errors like these when I run next build:
`Error occurred prerendering page "/admin/home". Read more: https://nextjs.org/docs/messages/prerender-error`
and 
`Error: NextRouter was not mounted. https://nextjs.org/docs/messages/next-router-not-mounted`

**Comment by isa-leroux448** [comment]
Is this a problem with the other PRs? Wondering if it's something I should fix in my branch or another bug fix ticket altogether

**Comment by AllanT102** [comment]
> I fixed the userInfo component used in the attendee information popup to use the new Registration type and be able to handle the nested user profile information. Right now I'm running into issues when building with next but I don't think they're related to my branch. I think some broken stuff got merged into dev because the package.lock syntax is off and I get a whole bunch of errors like these when I run next build:

please fix alongside this PR so it passes build, thanks

**Comment by voctory** [comment]
Any updates @isa-leroux448?

**Comment by isa-leroux448** [comment]
Sorry I forgot to request review again after Kevin fixed it. I've just resolved merge conflicts related to the sortable table columns. I've tested the table sorting and I think everything is working properly but might be worth a quick double-check on your end to make sure I didn't break anything.

**Comment by isa-leroux448** [comment]
@voctory bump that the PR is ready to review/merge.

File: src/pages/admin/event/[eventId]/[year]/stats/index.tsx [hunk 3]
```diff
@@ -2,39 +2,87 @@ import { AttendeeBasicInformation } from "@/types";
...
+      counts[String(value)] = counts[String(value)] + 1 || 1;
+    }
+  }
+  return Object.entries(counts)
+    .sort()
+    .map(([label, value]) => ({ label, value }));
+};
 
+const getNestedValue = (userData: Record<string, any>, field: string) => {
+  const keys = field.split(".");
```
Comment (voctory) [inline src/pages/admin/event/[eventId]/[year]/stats/index.tsx:31]: What's the format of `field`? Wondering why `field.split(".")` is needed

**Review by voctory — APPROVED** [review]
I left one comment about the `getNestedValue` function if you could take a look, otherwise LGTM

**Pull request merged**